x-pack/plugins/security/server/lib/authentication/providers/saml.ts

```diff
@@ -0,0 +1,557 @@
+    this.options.log(
+      ['debug', 'security', 'saml'],
+      `Trying to authenticate user request to ${request.url.path}.`
+    );
+
+    let {
+      authenticationResult,
+      // tslint:disable-next-line prefer-const
+      headerNotRecognized,
+    } = await this.authenticateViaHeader(request);
+    if (headerNotRecognized) {
+      return authenticationResult;
+    }
+
+    if (state && authenticationResult.notHandled()) {
+      authenticationResult = await this.authenticateViaState(request, state as SessionState);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = await this.authenticateViaRefreshToken(
+          request,
+          state as SessionState
+        );
+      }
+    }
+
+    if (authenticationResult.notHandled()) {
+      // Let's check if user is redirected to Kibana from IdP with valid SAMLResponse.
+      authenticationResult = await this.authenticateViaPayload(request, state as HandshakeState);
+    }
+
+    // If we couldn't authenticate by means of all methods above, let's try to
+    // initiate SAML handshake, otherwise just return authentication result we have.
+    return authenticationResult.notHandled()
+      ? await this.authenticateViaHandshake(request)
+      : authenticationResult;
+  }
+
+  /**
+   * Invalidates SAML access token if it exists.
+   * @param request HapiJS request instance.
+   * @param state State value previously stored by the provider.
+   */
+  public async deauthenticate(request: Request, state?: SessionState) {
+    this.options.log(
+      ['debug', 'security', 'saml'],
+      `Trying to deauthenticate user via ${request.url.path}.`
+    );
+
+    if ((!state || !state.accessToken) && !isSAMLRequestQuery(request.query)) {
+      this.options.log(
+        ['debug', 'security', 'saml'],
+        'There is neither access token nor SAML session to invalidate.'
+      );
+      return DeauthenticationResult.notHandled();
+    }
+
+    let logoutArgs: [
+      'shield.samlInvalidate' | 'shield.samlLogout',
+      { body: Record<string, string> }
+    ];
+    if (isSAMLRequestQuery(request.query)) {
+      this.options.log(
+        ['debug', 'security', 'saml'],
+        'Logout has been initiated by the Identity Provider.'
+      );
+      logoutArgs = [
+        'shield.samlInvalidate',
+        // Elasticsearch expects `queryString` without leading `?`, so we should strip it with `slice`.
+        {
+          body: {
+            queryString: request.url.search ? request.url.search.slice(1) : '',
+            acs: this.getACS(),
+          },
+        },
+      ];
+    } else {
+      this.options.log(['debug', 'security', 'saml'], 'Logout has been initiated by the user.');
+      logoutArgs = [
+        'shield.samlLogout',
+        { body: { token: state!.accessToken, refresh_token: state!.refreshToken } },
```

Comment:
I haven't seen any errors from the non-typescript code, but based on the types and the logic of this method, it seems like there's potential for a `TypeError` here if `state` is `undefined` and the `request.query.SAMLRequest` is specified.

Reply:
Hmm, am I missing something here?
* if both `state` and `request.query.SAMLRequest` are specified we'll enter line `197`
* if both aren't specified we'll enter line `185`
* if `state` is not specified, but `request.query.SAMLRequest` is we'll enter line `197`
* if `state` is specified, but `request.query.SAMLRequest` isn't we'll enter line `212`

Reply:
You're right, I missed that we were using `isSAMLRequestQuery` in both of the `if` statements on lines 185 and 197 and the impact this has. Ignore me!

Reply:
That's clearly an indicator of a wrongly organized code in this method so that both you and TS compiler (the reason I have these `state!.*`) are confused :) I'll see what I can do to make it less confusing in the follow-ups.